tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/WindowAggExec.java

```diff
@@ -347,6 +350,372 @@ private void finalizeWindow() {
+    int sameStartRange = 0; int sameEndRange = 0;
+    int frameStart = 0, frameEnd = accumulatedInTuples.size() - 1;
+    int actualStart = 0;
+
+    for (int i = 0; i < accumulatedInTuples.size(); i++) {
+      switch(windowFrameType) {
+        case TO_CURRENT_ROW:
+          frameEnd = i + windowFrameEndOffset; break;
+        case FROM_CURRENT_ROW:
+          frameStart = i + windowFrameStartOffset; break;
+        case SLIDING_WINDOW:
+          frameStart = i + windowFrameStartOffset;
+          frameEnd = i + windowFrameEndOffset;
+          break;
+      }
+
+      // RANGE window frame SHOULD include all the rows that has the same order by value with the current row
+      //   if comp == null, there is no order by and window frame is set as the entire partition
+      if (comp != null && windowFrameUnit == WindowSpec.WindowFrameUnit.RANGE) {
+        // move frame end point to the last row of the same order by value
+        if (sameEndRange == 0) {
+          sameEndRange = numOfTuplesWithTheSameKeyValue(frameEnd, comp, true);
+        }
+        sameEndRange --;
+        frameEnd += sameEndRange;
+
+        // move frame start point to the first row of the same order by value
+        if (sameStartRange == 0) {
+          sameStartRange = numOfTuplesWithTheSameKeyValue(frameStart, comp, true);
+          actualStart = frameStart;
+        }
+        sameStartRange --;
+        frameStart = actualStart;
+      }
+      // As the number of built-in window functions that support window frame is small,
+      // special treatment for each function seems to be reasonable
+      Tuple inTuple = getFunctionSpecificInput(funcName, frameStart, frameEnd);
+      Datum result = NullDatum.get();
+
+      if (inTuple != null) {
+        functions[functionIndex].merge(contexts[functionIndex], inTuple);
+        result = functions[functionIndex].terminate(contexts[functionIndex]);
+      }
+      Tuple outTuple = evaluatedTuples.get(i);
+      outTuple.put(nonFunctionColumnNum + functionIndex, result);
+    }
+  }
+
+  // it returns the number of Tuples with the same order by key value
+  // if only one Tuple for the given order by key exists, it returns 1
+  private int numOfTuplesWithTheSameKeyValue(int startOffset, BaseTupleComparator comp, boolean isForward) {
+    int numberOfTuples = 0;
+
+    Tuple inTuple = accumulatedInTuples.get(startOffset);
+    if (isForward) {
+      do {
+        numberOfTuples++;
+        startOffset++;
+      } while (startOffset < accumulatedInTuples.size() && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    } else {      // backward direction
+      do {
+        numberOfTuples++;
+        startOffset--;
+      } while (startOffset >= 0 && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    }
+
+    return numberOfTuples;
+  }
+
+  private Tuple getFunctionSpecificInput(String funcName, int dataStart, int dataEnd) {
```

Comment:
I have several reasons why I separately handle the built-in window functions with window frame support.
At first, there are only three built-in window functions that support frame, which are first_value, last_value, and nth_value. As the number of functions are fixed and small, I think code generality is less important than simple code.
Second, those functions commonly work on the position of window frame (start, end, or n-th row from start) so that above code does make much simpler to implement WindowAggExec, which now focuses on supporting window frame for aggregation functions only.

I agree that you have a point there.
If someone adds more built-in window functions with frame support, 
she or he may feel confused from different manner of implementation.
However, as I know, there are only one unimplemented function, nth_value, which can be easily added like first_value or last_value.